### DISCUSSION TITLE
`uvicorn.protocols.utils.get_client_addr` raises `TypeError: not enough arguments for format string`

### DISCUSSION THREAD

**Discussion by ducminh-phan** [Kludex/uvicorn#2195] [Potential Issue] [answered]
Minimal setup:

```python
from starlette.applications import Starlette
from starlette.requests import Request
from starlette.responses import JSONResponse
from starlette.routing import Route
from starlette.testclient import TestClient
from uvicorn.protocols.utils import get_client_addr


class Middleware:
    def __init__(self, app):
        self.app = app

    async def __call__(self, scope, receive, send):
        if scope["type"] == "http":
            client_addr = get_client_addr(scope)
            print(f"{client_addr=}")

        await self.app(scope, receive, send)


async def index(request: Request):
    return JSONResponse({})


app = Starlette(
    routes=[
        Route("/", index),
    ]
)
app.add_middleware(Middleware)


if __name__ == '__main__':
    test_client = TestClient(app)
    response = test_client.get("/")
    assert response.status_code == 200
```

Dependencies

```
asgiref==3.7.2
httpx==0.25.2
starlette==0.34.0
uvicorn==0.24.0.post1
```

Then run `pytest main.py`, the test failed due to the following error

```python
  File "/.../main.py", line 15, in __call__
    client_addr = get_client_addr(scope)
  File "/.../venv/lib/python3.9/site-packages/uvicorn/protocols/utils.py", line 46, in get_client_addr
    return "%s:%d" % client
TypeError: not enough arguments for format string

```

After debugging, I got the following information
* `starlette`'s `TestClient` sets the `client` key in `scope` to `["testclient", 50000]`: https://github.com/encode/starlette/blob/23c81da94b57701eabd43f582093442e6811f81d/starlette/testclient.py#L248.
* Then the line `"%s:%d" % client` fails because `client` is `["testclient", 50000]`, it would works if `client` is a tuple `("testclient", 50000)` instead of a list.

`asgiref`'s `typing` module states that `client` should be `Optional[Tuple[str, int]]`: https://github.com/django/asgiref/blob/d95e378e8bf5f0c8c87f4cfd888a0ac26c13fc5a/asgiref/typing.py#L77. So looks like `TestClient` incorrectly set the type of `scope["client"]` (`scope["server"]` as well).

However, [the ASGI spec](https://asgi.readthedocs.io/en/latest/specs/www.html#http-connection-scope) states that
> client (Iterable[Unicode string, int]) – A two-item iterable of [host, port], where host is the remote host’s IPv4 or IPv6 address, and port is the remote port as an integer. Optional; if missing defaults to None.

So a list of two elements is a valid value for `scope["client"]`. How should we resolve this discrepancy?

**Comment by Kludex** [comment, marked as answer]
The `get_client_addr` is not meant to be used directly, but Uvicorn's behavior is fine.
